src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -178,14 +181,87 @@ class SetupPyChrootRequest:
     py2: bool  # Whether to use py2 or py3 package semantics.
 
 
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class SetupKwargs:
+    """The keyword arguments to the `setup()` function in the generated `setup.py`."""
+
+    _pickled_bytes: bytes
+
+    def __init__(
+        self, kwargs: Mapping[str, Any], *, address: Address, _allow_banned_keys: bool = False
+    ) -> None:
+        super().__init__()
+        if "version" not in kwargs:
+            raise ValueError(f"Missing a `version` kwarg in the `provides` field for {address}.")
+
+        if not _allow_banned_keys:
+            for arg in {"data_files", "package_dir", "package_data", "packages"}:
+                if arg in kwargs:
+                    raise ValueError(
+                        f"{arg} cannot be set in the `provides` field for {address}, but it was "
+                        f"set to {kwargs[arg]}. Pants will dynamically set the value for you."
+                    )
+
+        # We serialize with `pickle` so that is hashable. We don't use `FrozenDict` because it
+        # would require that all values are immutable, and we may have lists and dictionaries as
+        # values. It's too difficult/clunky to convert those all, then to convert them back out of
+        # `FrozenDict`. We don't use JSON because it does not preserve data types like `tuple`.
+        self._pickled_bytes = pickle.dumps({k: v for k, v in sorted(kwargs.items())})
```

Comment:
How sure are we that it's OK to cache picked data?  For example, since you don't specify the protocol, this will use `DEFAULT_PROTOCOL`, which is different in different versions of Python. It will work fine now because in practice the possible protocols exist in all python versions that Pants runs on, but that might not hold true in the future. So you could end up writing a version that another process running on an earlier interpreter cannot read.

Reply:
Good point. https://github.com/pantsbuild/pants/pull/10725 sets the protocol to always be 4.